flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/join/lookup/AsyncLookupJoinRunner.java

```diff
@@ -0,0 +1,274 @@
+
+	public AsyncLookupJoinRunner(
+			GeneratedFunction<AsyncFunction<BaseRow, Object>> generatedFetcher,
+			GeneratedResultFuture<TableFunctionResultFuture<BaseRow>> generatedResultFuture,
+			TypeInformation<?> fetcherReturnType,
+			BaseRowTypeInfo rightRowTypeInfo,
+			boolean isLeftOuterJoin,
+			int asyncBufferCapacity) {
+		this.generatedFetcher = generatedFetcher;
+		this.generatedResultFuture = generatedResultFuture;
+		this.isLeftOuterJoin = isLeftOuterJoin;
+		this.asyncBufferCapacity = asyncBufferCapacity;
+		this.fetcherReturnType = fetcherReturnType;
+		this.rightRowTypeInfo = rightRowTypeInfo;
+	}
+
+	@Override
+	public void open(Configuration parameters) throws Exception {
+		super.open(parameters);
+		this.fetcher = generatedFetcher.newInstance(getRuntimeContext().getUserCodeClassLoader());
+		FunctionUtils.setFunctionRuntimeContext(fetcher, getRuntimeContext());
+		FunctionUtils.openFunction(fetcher, parameters);
+
+		// try to compile the generated ResultFuture, fail fast if the code is corrupt.
+		generatedResultFuture.compile(getRuntimeContext().getUserCodeClassLoader());
+
+		// row converter is stateless which is thread-safe
+		RowConverter rowConverter;
+		if (fetcherReturnType instanceof RowTypeInfo) {
+			rowConverter = (RowConverter) DataFormatConverters.getConverterForTypeInfo(fetcherReturnType);
+		} else if (fetcherReturnType instanceof BaseRowTypeInfo) {
+			rowConverter = null;
+		} else {
+			throw new IllegalStateException("This should never happen, " +
+				"currently fetcherReturnType can only be BaseRowTypeInfo or RowTypeInfo");
+		}
+
+		// asyncBufferCapacity + 1 as the queue size in order to avoid
+		// blocking on the queue when taking a collector.
+		this.resultFutureBuffer = new ArrayBlockingQueue<>(asyncBufferCapacity + 1);
+		for (int i = 0; i < asyncBufferCapacity + 1; i++) {
+			JoinedRowResultFuture rf = new JoinedRowResultFuture(
+				resultFutureBuffer,
+				createFetcherResultFuture(),
+				rowConverter,
+				isLeftOuterJoin,
+				rightRowTypeInfo.getArity());
+			// add will throw exception immediately if the queue is full which should never happen
+			resultFutureBuffer.add(rf);
+		}
+	}
+
+	@Override
+	public void asyncInvoke(BaseRow input, ResultFuture<BaseRow> resultFuture) throws Exception {
+		JoinedRowResultFuture outResultFuture = resultFutureBuffer.take();
+		// the input row is copied when object reuse in AsyncWaitOperator
+		outResultFuture.reset(input, resultFuture);
+
+		// fetcher has copied the input field when object reuse is enabled
+		fetcher.asyncInvoke(input, outResultFuture);
+	}
+
+	public TableFunctionResultFuture<BaseRow> createFetcherResultFuture() throws Exception {
+		TableFunctionResultFuture<BaseRow> resultFuture = generatedResultFuture.newInstance(
+			getRuntimeContext().getUserCodeClassLoader());
+		FunctionUtils.setFunctionRuntimeContext(resultFuture, getRuntimeContext());
+		FunctionUtils.openFunction(resultFuture, new Configuration());
```

Comment:
Consider saving `parameters` of `open` for use here?

Reply:
Sure